### ISSUE TITLE
Request for Position: Document Picture in Picture API

### ISSUE THREAD

**Issue by liberato-at-chromium** [mozilla/standards-positions#670]
## Request for Mozilla Position on an Emerging Web Specification

* Specification Title: Document Picture-in-Picture
* Explainer URL: https://github.com/steimelchrome/document-pip-explainer/blob/main/explainer.md
* Caniuse.com URL (optional): 
* Bugzilla URL (optional): 
* Mozillians who can provide input (optional): 

### Other information

The API that's in the explainer is approximate but well within the spirit of what we'd like to end up with.  Any feedback is appreciated! 

Link to chromestatus entry: https://chromestatus.com/feature/5755179560337408

**Comment by nschonni** [comment]
Is this different than the previous one in https://github.com/mozilla/standards-positions/issues/72

**Comment by liberato-at-chromium** [comment]
Yes, it is different.  #72 is for video content only.  This newer proposal provides support for arbitrary web content in a picture in picture window.

**Comment by muratcorlu** [comment]
This remembers me old-school popups. PiP for specific type of content (for video only for now) is a safe approach for most of the users. But if websites got access for putting any content to a PiP window, I'm afraid we'll see very annoying advertisement windows everywhere. And people will start searching for "how to disable PiP".

**Comment by bgrins** [comment]
I'd like to hear more about how UAs should expect to handle sites overusing / abusing this - particularly with a new windowing mode that may not be perceived as coming from the web page. Should popup blockers treat this identically to window.open? And would those controls affect video PiP as well?

**Comment by liberato-at-chromium** [comment]
> This reminds me of old-school popups.

Thank you for voicing these concerns. I agree that popups would not be a good outcome.  However, I don't believe that Document PiP makes the situation any worse.

Specifically, even with video-only PiP, sites can display arbitrary content; it's easy to make a one-frame video and/or redirect a video stream from a `<canvas>` to a PiP window for live, site-controlled content.

Further, the proposal for Document PiP is that it requires a user gesture, which is the same as video-only PiP or (usually, but I believe not required by spec) `window.open()`.  For a site to open a Document PiP window, it would already be able to open either of the other types anyway.  Plus, the lifetime of the PiP window is explicitly tied to that of the opener, unlike `window.open()`; when the opener closes / navigates away, any PiP window closes.  This is the same as video-only PiP.

The value in this proposal, as I see it, is allowing sites to provide more value via input.

Since these questions are in the same spirit as the above, I'll answer them together:

> Should popup blockers treat this identically to window.open? And would those controls affect video PiP as well?

We want to expose events that extensions could watch for to allow blocking.

> particularly with a new windowing mode that may not be perceived as coming from the web page

Making sure that it's clearly web content is definitely a goal, though it'll be up to the UA exactly how to do that.  For example, we plan to have Chromium display the origin of the opener in a slimmed-down title bar so that it's clear to the user what they're interacting with.

**Comment by abflow** [comment]
This is an essential feature for modern multi-windows web application.

**Comment by foolip** [comment]
FYI, there's an Intent to Ship for this feature on blink-dev now: https://groups.google.com/a/chromium.org/g/blink-dev/c/JTPl7fM64Lc/m/t-qwrAKmAAAJ

**Comment by bllngr** [comment]
And it's now available in Google Meet on Chrome browsers: https://workspaceupdates.googleblog.com/2023/06/enhanced-experience-picture-in-picture-google-meet-chrome.html

**Comment by steimelchrome** [comment]
For the initial launch of this API, Chrome disabled the ability for websites to use the resizeTo() and resizeBy() Window APIs to avoid spammy abuse given the always-on-top nature of a document picture-in-picture window. We've received feedback from multiple websites that having access to these APIs would be useful (e.g. clicking to expand a playlist). We propose a change to allow access to those APIs, but gated behind a user gesture (consumed) to limit potential abuse.

See the PR here: https://github.com/WICG/document-picture-in-picture/pull/104

**Comment by beaufortfrancois** [comment]
As [requested](https://github.com/WICG/document-picture-in-picture/issues/87#issuecomment-1940763854) by developers, we proposed adding [display-mode for picture-in-picture](https://drafts.csswg.org/mediaqueries-5/#display-mode-picture-in-picture) to CSS Media Queries Level 5.

```css
@media all and (display-mode: picture-in-picture) {
  body {
    margin: 0;
  }
  h1 {
    font-size: 0.8em;
  }
}
```
See the PR here: https://github.com/w3c/csswg-drafts/pull/9920

**Comment by steimelchrome** [comment]
Another addition we're proposing is a new boolean parameter allowReturnToTab, which defaults to true. When set to false, it hints to the user agent that showing a button in the document picture-in-picture UI that allows the user to return to the opener tab does not make sense for their use case, so the user agent can hide the button.

Initial request: https://github.com/WICG/document-picture-in-picture/issues/113
PR: https://github.com/WICG/document-picture-in-picture/pull/114

**Comment by steimelchrome** [comment]
We're also proposing allowing user gestures in the document picture-in-picture window to be usable in the opener window and vice versa. This makes it more ergonomic to use user-activation-gated APIs, since often event handlers in the document picture-in-picture window are actually run in the opener's context, so the opener's context needs access to the user gesture. This essentially makes the document picture-in-picture window act the same as a same-origin iframe inside the opener as far as user gesture propagation is concerned.

PR: https://github.com/WICG/document-picture-in-picture/pull/117

**Comment by beaufortfrancois** [comment]
For info, Spotify folks are using the Document Picture-in-Picture API for their Miniplayer.
You can learn more about their journey and use cases at https://developer.chrome.com/blog/spotify-picture-in-picture

**Comment by steimelchrome** [comment]
Adding a small change to copy the Document mode from the opener (e.g. quirks mode or non-quirks mode):

PR: https://github.com/WICG/document-picture-in-picture/pull/128
Chromestatus: https://chromestatus.com/feature/6275022143094784

**Comment by projektorius96** [comment]
> For info, Spotify folks are using the Document Picture-in-Picture API for their Miniplayer. You can learn more about their journey and use cases at https://developer.chrome.com/blog/spotify-picture-in-picture

[As per discussion](https://github.com/WICG/document-picture-in-picture/issues/121), would hiding the origin in so-called omnibox would be ever considered by modern browser's implementers ? Also could we, mundane developers expect, if not in Firefox , at least then in Chrome to see one day the support of window control overlay as found in PWA case ? It would be nice to see PWA like features exposed to PiP instantiated window as currently we're restricted only to user-agent native defaults, e.g. Chrome by default gives you "black" title bar and this is not bound to Chrome's default theme setting exposed to end-user.

**Comment by smaug----** [comment]
There are definitely some issues in the spec proposal, or Chrome's implementation not matching the spec, like apparently PiP affects fullscreen API, but it isn't defined how. 

I think conceptually this is quite reasonable. It is not that different from window.open("about:blank", "", "popup");

@steimelchrome are there plans to match the spec and Chrome's implementation?

**Comment by smaug----** [comment]
Hmm, mobile story is something to figure out. [Looks like](https://mozilla.pettay.fi/moztests/dip.html) Chrome doesn't support the API on Android.

**Comment by steimelchrome** [comment]
Yes ideally the spec and Chrome's implementation would match. That's a good point that the spec doesn't mention fullscreen, I'll open an issue for that. In regards to resizeTo/By(), I believe https://wicg.github.io/document-picture-in-picture/#resizing-the-pip-window defines it reasonably well, but if you think there's something we should say differently in that section, I'm definitely open to improvements

**Comment by smaug----** [comment]
Oh, I see, it was the a bit unusual language in the proposal, "user gesture" meaning basically transient activation.

**Comment by projektorius96** [comment]
> > For info, Spotify folks are using the Document Picture-in-Picture API for their Miniplayer. You can learn more about their journey and use cases at https://developer.chrome.com/blog/spotify-picture-in-picture
> 
> [As per discussion](https://github.com/WICG/document-picture-in-picture/issues/121), would hiding the origin in so-called omnibox would be ever considered by modern browser's implementers ? Also could we, mundane developers expect, if not in Firefox , at least then in Chrome to see one day the support for window control overlay as found in PWA case ? It would be nice to see PWA like features exposed to PiP instantiated window as currently we're restricted only to user-agent native defaults, e.g. Chrome by default gives you "black" title bar and this is not bound to Chrome's default theme setting exposed to end-user...

Just wanted to ask for support regarding [this feature request](https://issues.chromium.org/issues/392073250). If anyone think hiding omnibox (non-navigable address bar), please upvote. Ideally once implemented with [Borderless mode](https://issues.chromium.org/issues/40781978), PiP could look like this (see screenshot below)

<img width="959" alt="Image" src="https://github.com/user-attachments/assets/e82968a3-3dbb-43e5-98e8-dac4d3786928" />

**Comment by zcorpan** [comment]
Adding `concerns: device independence` because this is desktop-only in Chrome and https://github.com/WebKit/standards-positions/issues/41#issuecomment-1272177801 suggests it might not be implementable on iOS currently. Though in theory the PiP concept should work on mobile.

**Comment by DimDimShishkov** [comment]
@zcorpan , Could you clarify what the closing of this request indicates? Does it mean that the implementation of this mechanism will not be considered in the near future, or that it has actually been taken into active work?

**Comment by vinhill** [comment]
We're actively working on this in https://bugzilla.mozilla.org/show_bug.cgi?id=1858562
